fix(snuba): Add appropriate `UseCaseKey` for indexer [TET-146]

andriisoldatenko opened the pull request:
Update indexer invocation call to have the appropriate
`UseCaseKey` depending on use case.

In `src/sentry/sentry_metrics/indexer/base.py::StringIndexer`
when using `resolve` and `reverse_resolve` callers should not
rely on the default use_case_id.

Important changes:
- Add required parameter `use_case_id: UseCaseKey` to `get_series` from `src/sentry/snuba/metrics/datasource.py#L612`;
- Add required parameter  to `get_metrics` in `src/sentry/snuba/metrics/datasource.py`
- Add required parameter  to `get_tags` in `src/sentry/snuba/metrics/datasource.py`
- Add required parameter to `get_tag_values` in `src/sentry/snuba/metrics/datasource.py`
<!-- Describe your PR here. -->

Commit:
fix(snuba): Add appropriate `UseCaseKey` for indexer

Update indexer invocation call to have the appropriate
`UseCaseKey` depending on use case.

In `src/sentry/sentry_metrics/indexer/base.py::StringIndexer`
when using `resolve` and `reverse_resolve` callers should not
rely on the default use_case_id.

Commit:
fixup!

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fixup!

Commit:
update tests

Commit:
update tests

Commit:
update tests

Commit:
update tests

Commit:
revert back src/sentry/sentry_metrics/indexer/postgres_v2.py

Commit:
revert back src/sentry/sentry_metrics/indexer/postgres_v2.py

Commit:
revert src/sentry/utils/services.py

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

ahmedetefy commented on src/sentry/release_health/metrics.py:
```diff
@@ -376,7 +377,7 @@ def _count_users(total: bool, referrer: str) -> Dict[Any, int]:
 
         for project_id, release in project_releases:
-            release_tag_value = indexer.resolve(org_id, release)
+            release_tag_value = indexer.resolve(org_id, release, use_case_id=UseCaseKey.PERFORMANCE)
```
This module is `src/sentry/release_health/metrics.py`, why did you set it to `UseCaseKey.RELEASE_HEALTH`

andriisoldatenko replied:
```suggestion
            release_tag_value = indexer.resolve(org_id, release, use_case_id=UseCaseKey.RELEASE_HEALTH)
```

ahmedetefy commented on src/sentry/search/events/builder.py:
```diff
@@ -1822,7 +1823,9 @@ def resolve_snql_function(
-            result = indexer.resolve(self.organization_id, value)
+            result = indexer.resolve(
+                self.organization_id, value, use_case_id=UseCaseKey.RELEASE_HEALTH
+            )
```
This is discover code, should be performance usecase

andriisoldatenko replied:
@ahmedetefy i assume we also need to fix it also in https://github.com/getsentry/sentry/blob/master/tests/sentry/search/events/test_builder.py#L778-L785, because string has "transaction" metrics in this specific test.

andriisoldatenko commented on src/sentry/sentry_metrics/utils.py:
```diff
@@ -59,7 +60,7 @@ def resolve_weak(org_id: int, string: str) -> int:
     explicitly handling that exception.
     """
-    resolved = indexer.resolve(org_id, string)
+    resolved = indexer.resolve(org_id, string, use_case_id=UseCaseKey.RELEASE_HEALTH)
```
```suggestion
    resolved = indexer.resolve(org_id, string, use_case_id=UseCaseKey.PERFORMANCE)
```

ahmedetefy commented on src/sentry/snuba/metrics/datasource.py:
```diff
@@ -180,6 +181,7 @@ def get_custom_measurements(
     organization: Optional[Organization] = None,
     start: Optional[datetime] = None,
     end: Optional[datetime] = None,
+    use_case_id: UseCaseKey = UseCaseKey.RELEASE_HEALTH,
```
default here should be performance as this is custom transaction measurements

andriisoldatenko replied:
fixed.

ahmedetefy commented on tests/sentry/snuba/metrics/test_datasource.py:
```diff
@@ -101,7 +102,10 @@ def test_simple(self):
+            projects=[self.project],
+            organization=self.organization,
+            start=self.day_ago,
+            use_case_id=UseCaseKey.RELEASE_HEALTH,
```
performance use case

ahmedetefy commented on tests/sentry/snuba/metrics/test_datasource.py:
```diff
@@ -140,8 +144,12 @@ def test_metric_outside_query_daterange(self):
+            projects=[self.project],
+            organization=self.organization,
+            start=self.day_ago,
+            use_case_id=UseCaseKey.RELEASE_HEALTH,
```
performance here

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -255,10 +266,12 @@ def test_dist_count_aggregation_on_tx_satisfaction(self):
                             [
                                 Column(
-                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
+                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value, use_case_id=UseCaseKey.RELEASE_HEALTH)}]"
```
this is transaction, performance usecase

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -219,11 +229,12 @@ def test_set_count_aggregation_on_tx_satisfaction(self):
                             [
                                 Column(
-                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
+                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value, use_case_id=UseCaseKey.RELEASE_HEALTH)}]"
```
performance

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -255,10 +266,12 @@ def test_dist_count_aggregation_on_tx_satisfaction(self):
-                                    self.org_id, TransactionSatisfactionTagValue.SATISFIED.value
+                                    self.org_id,
+                                    TransactionSatisfactionTagValue.SATISFIED.value,
+                                    use_case_id=UseCaseKey.RELEASE_HEALTH,
```
performance

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -288,10 +301,12 @@ def test_dist_count_aggregation_on_tx_satisfaction(self):
                             [
                                 Column(
-                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
+                                    f"tags[{indexer.resolve(self.org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value, use_case_id=UseCaseKey.RELEASE_HEALTH)}]"
```
performance

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -288,10 +301,12 @@ def test_dist_count_aggregation_on_tx_satisfaction(self):
-                                    self.org_id, TransactionSatisfactionTagValue.TOLERATED.value
+                                    self.org_id,
+                                    TransactionSatisfactionTagValue.TOLERATED.value,
+                                    use_case_id=UseCaseKey.RELEASE_HEALTH,
```
performance

ahmedetefy reviewed (changes requested):
I would also expect that functions that we expose like `get_series` to accept some use case

andriisoldatenko committed:
Update src/sentry/release_health/metrics.py

andriisoldatenko committed:
Update src/sentry/sentry_metrics/utils.py

getsantry[bot] committed:
style(lint): Auto commit lint changes

getsantry[bot] committed:
style(lint): Auto commit lint changes

getsantry[bot] committed:
style(lint): Auto commit lint changes

Commit:
add use_case_id to get_series

Commit:
fix mypy linting

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix test

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix test

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix snuba tests

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
Add required parameter to get_metrics in src/sentry/snuba/metrics/datasource.py

Commit:
fix test

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix test

Commit:
fix tests/sentry/incidents/test_subscription_processor.py

Commit:
fix test

ahmedetefy commented on src/sentry/api/endpoints/organization_metrics.py:
```diff
@@ -28,7 +29,7 @@ def get(self, request: Request, organization) -> Response:
 
         projects = self.get_projects(request, organization)
-        metrics = get_metrics(projects)
+        metrics = get_metrics(projects, UseCaseKey.PERFORMANCE)
```
Why is this `UseCaseKey.Performance` ?

ahmedetefy commented on src/sentry/api/endpoints/organization_metrics.py:
```diff
@@ -69,7 +70,7 @@ def get(self, request: Request, organization) -> Response:
         projects = self.get_projects(request, organization)
         try:
-            tags = get_tags(projects, metric_names)
+            tags = get_tags(projects, metric_names, UseCaseKey.PERFORMANCE)
```
Same here why is this performance user case?

ahmedetefy commented on src/sentry/api/endpoints/organization_metrics.py:
```diff
@@ -123,7 +124,9 @@ def data_fn(offset: int, limit: int):
-                data = get_series(projects, query.to_metrics_query())
+                data = get_series(
+                    projects, query.to_metrics_query(), use_case_id=UseCaseKey.PERFORMANCE
+                )
```
This is our metrics API, why are all the calls with performance use case?

ahmedetefy commented on src/sentry/release_health/metrics.py:
```diff
@@ -376,7 +392,9 @@ def _count_users(total: bool, referrer: str) -> Dict[Any, int]:
         for project_id, release in project_releases:
-            release_tag_value = indexer.resolve(org_id, release)
+            release_tag_value = indexer.resolve(
+                org_id, release, use_case_id=UseCaseKey.RELEASE_HEALTH
```
Can we make `us_case_id` always the last arg like it is here but in other calls as well? Seems strange to me that this use_case_id is the first arg passed

ahmedetefy commented on src/sentry/release_health/metrics.py:
```diff
@@ -191,15 +192,19 @@ def _get_crash_free_rate_data(
+                Condition(
+                    Column("metric_id"),
+                    Op.EQ,
+                    resolve(UseCaseKey.RELEASE_HEALTH, org_id, SessionMRI.SESSION.value),
```
Nit: Since this is all a release health module I would just instantiate it in the top of the module and just reuse it 
`USE_CASE = UseCaseKey.RELEASE_HEALTH`

ahmedetefy commented on src/sentry/sentry_metrics/utils.py:
```diff
@@ -14,17 +15,17 @@ class MetricIndexNotFound(InvalidParams):  # type: ignore
 
 
-def reverse_resolve(index: int) -> str:
+def reverse_resolve(use_case_id: UseCaseKey, index: int) -> str:
```
To me it seems strange to have the `use_case_id` be the first argument, can you just make it the last?
In this entire module I mean

ahmedetefy commented on src/sentry/snuba/metrics/fields/snql.py:
```diff
@@ -110,9 +115,9 @@ def _snql_on_tx_satisfaction_factory(org_id, satisfaction_value: str, metric_ids
                             [
                                 Column(
-                                    f"tags[{resolve_weak(org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
+                                    f"tags[{resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
```
This says `TRANSACTION_SATISFACTION`, why is Usecase here release health

ahmedetefy commented on src/sentry/snuba/metrics/fields/snql.py:
```diff
@@ -110,9 +115,9 @@ def _snql_on_tx_satisfaction_factory(org_id, satisfaction_value: str, metric_ids
+                                    f"tags[{resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, TransactionTagsKey.TRANSACTION_SATISFACTION.value)}]"
                                 ),
-                                resolve_weak(org_id, satisfaction_value),
+                                resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, satisfaction_value),
```
Same here

ahmedetefy commented on tests/sentry/incidents/test_subscription_processor.py:
```diff
@@ -2134,7 +2135,13 @@ def test_ensure_case_when_no_metrics_index_not_found_is_handled_gracefully(self)
+                    "data": [
+                        {
+                            resolve_tag_key(
+                                UseCaseKey.PERFORMANCE, self.organization.id, "session.status"
```
This says `session.status`, why `UseCaseKey.Performance` ?

ahmedetefy commented on tests/sentry/incidents/test_subscription_processor.py:
```diff
@@ -2187,9 +2194,9 @@ def send_crash_rate_alert_update(self, rule, value, subscription, time_delta=Non
-            session_status = resolve_tag_key(org_id, "session.status")
-            tag_value_init = resolve_weak(org_id, "init")
-            tag_value_crashed = resolve_weak(org_id, "crashed")
+            session_status = resolve_tag_key(UseCaseKey.PERFORMANCE, org_id, "session.status")
```
This is release health use case

Commit:
addressed code review comments

Commit:
addressed code review comments

andriisoldatenko commented:
@ahmedetefy PTAL.

Commit:
one small missing rename

Commit:
fixup!

Commit:
fix tests

ahmedetefy commented on src/sentry/snuba/metrics/fields/base.py:
```diff
@@ -235,11 +241,21 @@ class AliasedDerivedMetric(AliasedDerivedMetricDefinition, MetricObject):
 
     def generate_metric_ids(self, projects: Sequence[Project]) -> Set[int]:
-        return {resolve_weak(org_id_from_projects(projects), self.raw_metric_mri)}
+        return {
```
same here, this module is generic. the use case needs to be passed down from get_series

ahmedetefy commented on src/sentry/snuba/metrics/fields/base.py:
```diff
@@ -235,11 +241,21 @@ class AliasedDerivedMetric(AliasedDerivedMetricDefinition, MetricObject):
+            Function(
+                "equals",
+                [
+                    Column("metric_id"),
```
same here, usecase need to be passed same way org_id is passed

untitaker replied:
wdym ahmed?

ahmedetefy replied:
Cause we chatted earlier that there was a similar PR that he could look to for inspo which Meredith added with org_id so this a reference to that

andriisoldatenko replied:
this fixed already here https://github.com/getsentry/sentry/blob/a2e7352dd2257e52cd3b5596d53f85932ef84ea3/src/sentry/snuba/metrics/fields/base.py#L252-L252

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -791,10 +796,10 @@ def test_orderby_percentile_with_many_fields_one_entity(self):
-        transaction_2 = _indexer_record(org_id, "/bar/")
+        metric_id_fcp = perf_indexer_record(org_id, TransactionMRI.MEASUREMENTS_FCP.value)
+        transaction_id = perf_indexer_record(org_id, "transaction")
+        transaction_1 = rh_indexer_record(org_id, "/foo/")
```
This says transaction, why are you using `rh_indexer_record`

andriisoldatenko replied:
same as in https://github.com/getsentry/sentry/pull/36308#discussion_r917724649

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -725,9 +730,9 @@ def test_limit_with_orderby_is_overridden_by_paginator_limit(self):
-        tag1 = _indexer_record(org_id, "tag1")
-        value1 = _indexer_record(org_id, "value1")
-        value2 = _indexer_record(org_id, "value2")
+        tag1 = rh_indexer_record(org_id, "tag1")
```
This is a transaction test, why are you using rh_indexer_record?

andriisoldatenko replied:
@ahmedetefy as i mentioned early - this test send GET request to `/api/0/organizations/baz/metrics/data/` and this endpoint has https://github.com/getsentry/sentry/blob/3717db2a928b5c807dd6986992b3b103941a2c45/src/sentry/api/endpoints/organization_metrics.py#L128 `RELEASE_HEALTH` param in `get_series` function call. So if i chnage to:

```diff
diff --git tests/sentry/api/endpoints/test_organization_metric_data.py tests/sentry/api/endpoints/test_organization_metric_data.py
index 64a72c5639..eecab25acc 100644
--- tests/sentry/api/endpoints/test_organization_metric_data.py
+++ tests/sentry/api/endpoints/test_organization_metric_data.py
@@ -730,9 +730,9 @@ class OrganizationMetricDataTest(MetricsAPIBaseTestCase):
         `limit` parameter
         """
         org_id = self.organization.id
-        tag1 = rh_indexer_record(org_id, "tag1")
-        value1 = rh_indexer_record(org_id, "value1")
-        value2 = rh_indexer_record(org_id, "value2")
+        tag1 = perf_indexer_record(org_id, "tag1")
+        value1 = perf_indexer_record(org_id, "value1")
+        value2 = perf_indexer_record(org_id, "value2")
```

test won't work.

Commit:
address 3rd round of code review

Commit:
fix mypy

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

ahmedetefy commented on src/sentry/snuba/metrics/fields/snql.py:
```diff
@@ -60,9 +63,11 @@ def _get_snql_conditions(org_id, metric_ids, exclude_tx_statuses):
 
         tx_col = Column(
-            f"tags[{resolve_weak(org_id, TransactionTagsKey.TRANSACTION_STATUS.value)}]"
+            f"tags[{resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, TransactionTagsKey.TRANSACTION_STATUS.value)}]"
```
Can you update this please as per the comment?

ahmedetefy replied:
@andriisoldatenko

andriisoldatenko replied:
one sec.

andriisoldatenko replied:
fixed.

ahmedetefy commented on src/sentry/snuba/metrics/fields/snql.py:
```diff
@@ -60,9 +63,11 @@ def _get_snql_conditions(org_id, metric_ids, exclude_tx_statuses):
         )
-        excluded_statuses = [resolve_weak(org_id, s) for s in exclude_tx_statuses]
+        excluded_statuses = [
+            resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, s) for s in exclude_tx_statuses
```
Same here @andriisoldatenko

andriisoldatenko replied:
fixed in a2e7352dd2

ahmedetefy commented on tests/sentry/incidents/test_subscription_processor.py:
```diff
@@ -2187,9 +2194,9 @@ def send_crash_rate_alert_update(self, rule, value, subscription, time_delta=Non
-            tag_value_init = resolve_weak(org_id, "init")
-            tag_value_crashed = resolve_weak(org_id, "crashed")
+            session_status = resolve_tag_key(UseCaseKey.RELEASE_HEALTH, org_id, "session.status")
+            tag_value_init = resolve_weak(UseCaseKey.PERFORMANCE, org_id, "init")
```
This is session status, please set that use case to release health and the same with following line

andriisoldatenko replied:
fixed in a2e7352dd2

Commit:
fix few more places

andriisoldatenko commented on src/sentry/snuba/metrics/fields/base.py:
```diff
@@ -625,7 +641,7 @@ def __recursively_generate_metric_ids(cls, org_id: int, derived_metric_mri: str)
         for metric_mri in derived_metric.metrics:
             if metric_mri not in DERIVED_METRICS:
-                ids.add(resolve_weak(org_id, metric_mri))
+                ids.add(resolve_weak(UseCaseKey.RELEASE_HEALTH, org_id, metric_mri))
```
fixed in a2e7352dd2

andriisoldatenko commented on tests/sentry/snuba/metrics/test_query_builder.py:
```diff
@@ -855,7 +938,9 @@ def test_translate_meta_results():
     ]
-    assert translate_meta_results(meta, {"p50(transaction.measurements.lcp)"}) == sorted(
+    assert translate_meta_results(
+        meta, {"p50(transaction.measurements.lcp)"}, UseCaseKey.RELEASE_HEALTH
```
fixed in a2e7352dd2

andriisoldatenko commented:
This PR removes defaults https://github.com/getsentry/sentry/pull/36501

untitaker commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -22,8 +23,12 @@
+    return indexer.record(use_case_id=use_case_id, org_id=org_id, string=string)
+
+
+perf_indexer_record = partial(indexer_record, UseCaseKey.PERFORMANCE)
```
So if we make the metrics API only work on releasehealth, don't we always want to go to that usecase in this entire file?

andriisoldatenko replied:
correct, that was my assumption too, when @ahmedetefy suggested that this API endpoint is RELEASE_HEALTH only.
https://github.com/getsentry/sentry/blob/a2e7352dd2257e52cd3b5596d53f85932ef84ea3/src/sentry/api/endpoints/organization_metrics.py#L104-L143
Otherwise we need to expose it to query parameter.

untitaker replied:
i think either way makes sense. i'll leave the decision up to you. i don't think mixing the usecases in tests like a ton of sense, but since it's just tests and it all goes to the same DB in CI i guess it doesn't matter

untitaker reviewed (approved):
i think there's maybe an open question about what to do with metrics API. I think it makes sense to make it only work with release health if that's all we use it for, it's simpler for now.

generally this is fine and I would merge as-is. I also find the argument order of some functions confusing now but I think we can fix that later if we feel like it. Since I'm blocked on this being merged I'd rather merge it and do cosmetic changes later.

Commit:
add useCase optional query param

Commit:
add useCase optional query param

Commit:
fix type hints

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
add small TODO to not mix release_health and performance metrics in single unit test

ahmedetefy commented on src/sentry/sentry_metrics/configuration.py:
```diff
@@ -9,6 +9,15 @@ class UseCaseKey(Enum):
+        elif use_case in ("release-health", "releaseHealth"):
+            return UseCaseKey.RELEASE_HEALTH
+        else:
+            return UseCaseKey.RELEASE_HEALTH
```
Do we want to default in the `else` here? or should we raise an exception if it is not one of those use cases?

andriisoldatenko replied:
yeah originally i had exception, but since it's optional ... i'm not 100% sure what we want here. Do you want me to raise `NotImplementedError` or something more specific?

andriisoldatenko replied:
@ahmedetefy i decided to add raise exception: pls check it here 104af64353

Commit:
NotImplementedError for other that release-health and performance in UseCaseKey

wmak reviewed (approved):
Changes to visibility code lgtm

ahmedetefy commented on src/sentry/sentry_metrics/configuration.py:
```diff
@@ -9,6 +9,15 @@ class UseCaseKey(Enum):
+        elif use_case in ("release-health", "releaseHealth"):
+            return UseCaseKey.RELEASE_HEALTH
+        else:
+            raise NotImplementedError
```
^Yeah that's what I had in mind @andriisoldatenko when I suggested this change

Commit:
return ParseError whe useCase param has wrong value; update tests

Commit:
return ParseError whe useCase param has wrong value; update tests

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

markstory commented on src/sentry/api/bases/organization.py:
```diff
@@ -366,6 +367,17 @@ def convert_args(self, request: Request, organization_slug, *args, **kwargs):
+    def get_use_case_id(self, request: Request):
+        """
+        Extracts useCase from request and validate it agains UseCaseKey enum type
+        if request param is omitted fallback to UseCaseKey.RELEATH_HEALTH
```
```suggestion
        if request param is omitted fallback to UseCaseKey.RELEASE_HEALTH
```

markstory commented on src/sentry/api/bases/organization.py:
```diff
@@ -366,6 +367,17 @@ def convert_args(self, request: Request, organization_slug, *args, **kwargs):
+        try:
+            return UseCaseKey.from_str(request.GET.get("useCase", "release-health"))
+        except ValueError:
+            raise ParseError(detail="Invalid useCase parameter.")
```
Is there a list of valid choices that we could tell the user to choose from?

andriisoldatenko replied:
yeah one moment.

andriisoldatenko replied:
@markstory PTAL a230df69c2

andriisoldatenko committed:
Update src/sentry/api/bases/organization.py

Co-authored-by: Mark Story <mark@mark-story.com>

Commit:
add valid options to ParseError and update tests

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix unit test

edwardgou-sentry committed:
feat(discover): lowers rate limit for events endpoint (#36566)

Sets rate limit for eventsv2 and events api to 25 concurrent and 25 in 1 second.

wmak committed:
feat(custom-measurements): Support custom measurements in mqb (#36158)

- This adds custom measurements support in the metric query builder
- This doesn't update the meta yet, so custom measurements is showing up
  as durations even though they could be something else
- This doesn't handle incompatible queries and CM yet, and will fallback
  to discover for now

Commit:
resolve conflicts

Commit:
fix mypy

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix: test_indexer.py

ahmedetefy commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -139,7 +140,7 @@ def validate_orderby(self) -> None:
             # Construct a metrics expression
             metric_field_obj = metric_object_factory(f.field.op, metric_mri)
-            entity = metric_field_obj.get_entity(self.project_ids)
+            entity = metric_field_obj.get_entity(self.project_ids, UseCaseKey.RELEASE_HEALTH)
```
This is not necessarily true because this could be also a query for performance, so you need to propagate here the usecase from the request calling this somehow

ahmedetefy replied:
I know I said we shouldn't add the use case here https://github.com/getsentry/sentry/pull/36308/files#diff-7da0855ce9cf928e32249842b19bdcec279a052ee9fcc9b9583f2ff3fb7fea33L80 but if there is no other way to propagate it then maybe just add it as a parameter here. I would spend sometime thinking how to propagate this down without adding it to class it though

ahmedetefy commented on src/sentry/snuba/metrics_performance.py:
```diff
@@ -13,11 +13,12 @@
 
 
-def resolve_tags(results: Any, query_definition: MetricsQueryBuilder) -> Any:
+def resolve_tags(results: Any, query_definition: MetricsQueryBuilder):
```
Why did you remove the return type annotation here?

ahmedetefy commented on src/sentry/testutils/cases.py:
```diff
@@ -1129,19 +1130,19 @@ def _push_metric(cls, session, type, key: SessionMRI, tags, value):
 
         def metric_id(key: SessionMRI):
```
The type here says `SessionMRI` but the use case is perfomance, which one should be changed?

ahmedetefy commented on src/sentry/testutils/cases.py:
```diff
@@ -1186,6 +1187,60 @@ def _send_buckets(cls, buckets, entity):
         )
 
 
+class SessionMetricsReleaseHealthTestCase(SessionMetricsTestCase):
```
This class is a duplication of lines 1130 to 1168, can we figure out a way to pass use case without duplicating the code?

ahmedetefy commented on src/sentry/utils/services.py:
```diff
@@ -124,6 +124,12 @@ def expose(self, context: MutableMapping[str, Any]) -> None:
         base_instance = base()
         for key in itertools.chain(base.__all__, ("validate", "setup")):
             if inspect.isroutine(getattr(base_instance, key)):
+                # def foo(*args, **kwargs):
```
Why is this commented code here?

ahmedetefy commented on tests/sentry/snuba/metrics/test_snql.py:
```diff
@@ -36,7 +36,7 @@ def setUp(self):
         self.metric_ids = [0, 1, 2]
         indexer.bulk_record(
-            use_case_id=UseCaseKey.RELEASE_HEALTH,
+            use_case_id=UseCaseKey.PERFORMANCE,
```
This use case is release health as your line https://github.com/getsentry/sentry/pull/36308/files#diff-cd6ed2f06582566553d69e6c5ab1136505be2633563bf27e63eb2f84ad047c05R79 says

andriisoldatenko replied:
okey, but i see different metrics in bulk_create -> https://github.com/getsentry/sentry/blob/d23fb690138b93944c9a410a3ba6b5e2ab7005e2/tests/sentry/snuba/metrics/test_snql.py#L49-L56

ahmedetefy replied:
Can we maybe separate the this `bulk_record` ? As an example 
```
    indexer.bulk_record(
            use_case_id=UseCaseKey.RELEASE_HEALTH,
            org_strings={
                self.org_id: [
                    "abnormal",
                    "crashed",
                    "errored_preaggr",
                    "errored",
                    "exited",
                    "init",
                    "session.status",
                ]
            },
        )
     indexer.bulk_record(
            use_case_id=UseCaseKey.PERFORMANCE,
            org_strings={
                self.org_id: [
                    TransactionSatisfactionTagValue.FRUSTRATED.value,
                    TransactionSatisfactionTagValue.SATISFIED.value,
                    TransactionSatisfactionTagValue.TOLERATED.value,
                    TransactionStatusTagValue.CANCELLED.value,
                    TransactionStatusTagValue.OK.value,
                    TransactionStatusTagValue.UNKNOWN.value,
                    TransactionTagsKey.TRANSACTION_SATISFACTION.value,
                    TransactionTagsKey.TRANSACTION_STATUS.value,
                ]
            },
        )
```

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -1689,21 +1722,21 @@ def setUp(self):
+        self.release_tag = rh_indexer_record(self.organization.id, "release")
+        self.tx_metric = rh_indexer_record(org_id, TransactionMRI.DURATION.value)
+        self.tx_status = rh_indexer_record(org_id, TransactionTagsKey.TRANSACTION_STATUS.value)
+        self.transaction_lcp_metric = rh_indexer_record(
             self.organization.id, TransactionMRI.MEASUREMENTS_LCP.value
         )
-        self.tx_satisfaction = _indexer_record(
+        self.tx_satisfaction = rh_indexer_record(
             self.organization.id, TransactionTagsKey.TRANSACTION_SATISFACTION.value
         )
-        self.tx_user_metric = _indexer_record(self.organization.id, TransactionMRI.USER.value)
+        self.tx_user_metric = rh_indexer_record(self.organization.id, TransactionMRI.USER.value)
```
These all need to be performance use case lines 1731 to 1739

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2672,7 +2709,7 @@ def test_failure_rate_transaction(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_status: _indexer_record(
+                        self.tx_status: rh_indexer_record(
```
performance usecase

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2686,7 +2723,7 @@ def test_failure_rate_transaction(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_status: _indexer_record(
+                        self.tx_status: rh_indexer_record(
```
performance usecase

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2714,7 +2751,7 @@ def test_failure_rate_transaction(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_status: _indexer_record(
+                        self.tx_status: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2803,7 +2840,7 @@ def test_apdex_transactions(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2817,7 +2854,7 @@ def test_apdex_transactions(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2831,7 +2868,7 @@ def test_apdex_transactions(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2863,7 +2900,7 @@ def test_miserable_users(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2877,7 +2914,7 @@ def test_miserable_users(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2909,7 +2946,7 @@ def test_user_misery(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2923,7 +2960,7 @@ def test_user_misery(self):
                     "timestamp": user_ts,
                     "tags": {
-                        self.tx_satisfaction: _indexer_record(
+                        self.tx_satisfaction: rh_indexer_record(
```
performance use case

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -2991,9 +3028,9 @@ def test_session_duration_derived_alias(self):
-        tag1 = _indexer_record(org_id, "tag1")
-        value1 = _indexer_record(org_id, "value1")
-        value2 = _indexer_record(org_id, "value2")
+        tag1 = rh_indexer_record(org_id, "tag1")
+        value1 = rh_indexer_record(org_id, "value1")
+        value2 = rh_indexer_record(org_id, "value2")
```
performance use case as this is a query over transaction lcp

ahmedetefy commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -3038,9 +3075,9 @@ def test_histogram(self):
-        tag1 = _indexer_record(org_id, "tag1")
-        value1 = _indexer_record(org_id, "value1")
-        value2 = _indexer_record(org_id, "value2")
+        tag1 = rh_indexer_record(org_id, "tag1")
+        value1 = rh_indexer_record(org_id, "value1")
+        value2 = rh_indexer_record(org_id, "value2")
```
performance use case as this is a query over transaction lcp

ahmedetefy commented:
I think that the changes requested are probably the last changes required to get this PR through + fixing failing tests

andriisoldatenko commented on src/sentry/sentry_metrics/utils.py:
```diff
@@ -14,17 +15,17 @@ class MetricIndexNotFound(InvalidParams):  # type: ignore
 
 
-def reverse_resolve(index: int) -> str:
+def reverse_resolve(use_case_id: UseCaseKey, index: int) -> str:
```
the idea is very simple, right after we merge this PR we can create follow up PR to remove default value from `resolve` and `reverse_resolve` and i'm going to make it in similar manner as `bulk_record`:

https://github.com/getsentry/sentry/blob/d23fb690138b93944c9a410a3ba6b5e2ab7005e2/src/sentry/sentry_metrics/indexer/postgres_v2.py#L47

and `record`:

https://github.com/getsentry/sentry/blob/d23fb690138b93944c9a410a3ba6b5e2ab7005e2/src/sentry/sentry_metrics/indexer/postgres_v2.py#L166-L166

Also when it's first required parameter you can easily use `functools.partitial`.

ahmedetefy replied:
Ah I think I understand, you are saying that it is currently inconsistent anyways, and once you remove the default in the follow up PR, you will move `use_case_id` to first parameter anyways in those functions so that they are all consistent with `use_case_id` as the first argument?

ahmedetefy replied:
Okay that's fine then. Can we maybe add a ToDo comment with that so we don't forget?

Commit:
refactored based on code review

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
fix bad formating

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

getsantry[bot] committed:
style(lint): Auto commit lint changes

getsantry[bot] committed:
style(lint): Auto commit lint changes

ahmedetefy commented on src/sentry/snuba/entity_subscription.py:
```diff
@@ -476,8 +477,46 @@ def _aggregate_query_results_v2(
 
-    def get_snql_extra_conditions(self) -> List[Condition]:
-        return [Condition(Column("metric_id"), Op.EQ, resolve(self.org_id, self.metric_key.value))]
+    def build_query_builder(
```
Should this be included in the PR?

andriisoldatenko replied:
yep, i'm aware of it. Let me push current version.

Commit:
fixup!

ahmedetefy commented on src/sentry/search/events/builder.py:
```diff
@@ -1835,7 +1836,12 @@ def resolve_snql_function(
         """Layer on top of the metric indexer so we'll only hit it at most once per value"""
         if value not in self._indexer_cache:
-            result = indexer.resolve(self.organization_id, value)
+            use_case_id = UseCaseKey.PERFORMANCE
```
Your comment is a bit vague, what do you mean?

andriisoldatenko replied:
I'll remove it of course, so what i'm trying to understand is -> we still can invoke this method `resolve_metric_index ` for both cases: `UseCaseKey.PERFORMANCE` and `UseCaseKey.RELEASE_HEALTH`.

So i'm trying to find a way to pass it somehow - or tweak tests.

ahmedetefy replied:
I don't think this module should invoke release health, why are you saying that? Does the alert code use this?

andriisoldatenko replied:
Correct, it was my first impression. But there few tests that are doing it from code.

andriisoldatenko replied:
removed comment from code.

andriisoldatenko replied:
if you checkout my branch locally and run: `pytest tests/sentry/snuba/test_tasks.py -k test_query_and_environment_users_metrics` you can see:
```
―――――――――――――――――――――――――――――――――――――――――――――――― BuildSnqlQueryTest.test_query_and_environment_users_metrics ――――――――――――――――――――――――――――――――――――――――――――――――
tests/sentry/snuba/test_tasks.py:1006: in test_query_and_environment_users_metrics
    self.run_test(
tests/sentry/snuba/test_tasks.py:456: in run_test
    query_builder = build_query_builder(
src/sentry/snuba/tasks.py:187: in build_query_builder
    return entity_subscription.build_query_builder(query, project_ids, environment, params)
src/sentry/snuba/entity_subscription.py:345: in build_query_builder
    qb = AlertMetricsQueryBuilder(
src/sentry/search/events/builder.py:2160: in __init__
    super().__init__(*args, **kwargs)
src/sentry/search/events/builder.py:1608: in __init__
    super().__init__(
src/sentry/search/events/builder.py:164: in __init__
    self.resolve_query(
src/sentry/search/events/builder.py:1641: in resolve_query
    self.where, self.having = self.resolve_conditions(
src/sentry/search/events/builder.py:312: in resolve_conditions
    where = self.resolve_where(parsed_terms)
src/sentry/search/events/builder.py:270: in resolve_where
    condition = self.format_search_filter(term)
src/sentry/search/events/builder.py:949: in format_search_filter
    converted_filter = self.convert_search_filter_to_condition(
src/sentry/search/events/builder.py:1007: in convert_search_filter_to_condition
    return converter(search_filter)
src/sentry/search/events/builder.py:1869: in _default_filter_converter
    value = self._resolve_tag_value(value)
src/sentry/search/events/builder.py:1850: in _resolve_tag_value
    raise InvalidSearchQuery("Tag value was not found")
E   sentry.exceptions.InvalidSearchQuery: Tag value was not found
```
then if you apply:
```diff
diff --git src/sentry/search/events/builder.py src/sentry/search/events/builder.py
index 314866d4fa..7b6f01f3e5 100644
--- src/sentry/search/events/builder.py
+++ src/sentry/search/events/builder.py
@@ -1836,7 +1836,7 @@ class MetricsQueryBuilder(QueryBuilder):
     def resolve_metric_index(self, value: str) -> Optional[int]:
         """Layer on top of the metric indexer so we'll only hit it at most once per value"""
         if value not in self._indexer_cache:
-            use_case_id = UseCaseKey.PERFORMANCE
+            use_case_id = UseCaseKey.RELEASE_HEALTH
             result = indexer.resolve(self.organization_id, value, use_case_id=use_case_id)
             self._indexer_cache[value] = result
```
and rerun test it works.

ahmedetefy replied:
Ah okay.. I understand now, in that case you probably want to check if this query builder is building a query for crash rate alerts which are session alerts and then set release health case. Maybe check with @wmak to help you navigate that part of the code?

wmak replied:
You should be able to check `self.dataset is Dataset.Metrics` for release health and `self.dataset is Dataset.PerformanceMetrics` for MEP

If you're feeling keen we could add a class property that does this as well


```
@property
def is_performance_metrics(self):
    return self.dataset is Dataset.PerformanceMetrics
```

andriisoldatenko replied:
thanks! yea i were thinking about similar approach.

andriisoldatenko replied:
@wmak one small quiestion:
as you can see in this data class:
```
@unique
class Dataset(Enum):
    Events = "events"
    Transactions = "transactions"
    Discover = "discover"
    Outcomes = "outcomes"
    OutcomesRaw = "outcomes_raw"
    Sessions = "sessions"
    # Actually Release Health
    Metrics = "metrics"
    PerformanceMetrics = "generic_metrics"
```
i'm not fully sure how do we want to distinguish all data class fields into 2 sub-groups for `indexer`:
- release_health and performance
---
Previously we discussed with @ahmedetefy that we have 2 groups: 
`transactions` are belongs to `UseCaseKey.PERFORMANCE` value for indexer and
`sessions` are belongs to `UseCaseKey.RELEASE_HEALTH`.

wmak replied:
You don't have to handle all the Datasets, the `MetricQueryBuilder`s only accept two:
https://github.com/getsentry/sentry/blob/7f6c5f9e89c0ceb49e2d89959ca4781cdd0942e1/src/sentry/search/events/builder.py#L1607

ahmedetefy commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -139,7 +140,13 @@ def validate_orderby(self) -> None:
+
+            # Find correct use_case_id based on metric_name
+            use_case_id = UseCaseKey.RELEASE_HEALTH
+            if f.field.metric_name.startswith("transaction"):
```
There is no guarantee that the public name won't change. The only guarantee exists on MRI not changing, and so please check against name space in MRI regex not on the metric name

andriisoldatenko replied:
fixed.

Commit:
remove my personal comment

ahmedetefy reviewed (approved):
Since I am going on vacation, unblocking "Requested Changes" merge block when I am away but please do not merge until addressing the comments mentioned and wouldn't be the worst idea to have someone else look at it once its ready

Commit:
improved how we validate metric

andriisoldatenko committed:
Merge branch 'master' into andrii/update-indexer-invocation-call

Commit:
add is_performance property to search/events/builder.py

Commit:
fixup!

mitsuhiko commented:
There seems to be one failing test still that needs addressing. Sounds like something might be pulling the wrong data now.

Commit:
fix tests
